sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidFilterUnnestRule.java

```diff
@@ -94,9 +95,15 @@ public static DruidProjectOnUnnestRule instance()
     public boolean matches(RelOptRuleCall call)
     {
       final Project rightP = call.rel(0);
-      final SqlKind rightProjectKind = rightP.getChildExps().get(0).getKind();
-      // allow rule to trigger only if there's a string CAST or numeric literal cast
-      return rightP.getProjects().size() == 1 && (rightProjectKind == SqlKind.CAST || rightProjectKind == SqlKind.LITERAL);
+      if (rightP.getChildExps().size() > 0) {
+        final SqlKind rightProjectKind = rightP.getChildExps().get(0).getKind();
+        final SqlTypeName rightType = rightP.getChildExps().get(0).getType().getSqlTypeName();
+        // allow rule to trigger only if there's a non-decimal CAST or numeric literal cast
+        return rightP.getProjects().size() == 1 && ((rightProjectKind == SqlKind.CAST
+                                                     && rightType != SqlTypeName.DECIMAL)
+                                                    || rightProjectKind == SqlKind.LITERAL);
```

Comment:
Why?  What's wrong with a decimal cast?

Reply:
Queries such as `select sum(c) col from druid.numfoo, unnest(mv_to_array(dim3)) as u(c)` will need to cast string to decimal that will appear as a logical project on top of the unnest rel. We do not want this rule to be triggered if this is such a case.

Reply:
Fixed this to allow only if we cast from same datatype to same datatype. We have tests covering for string, float etc